### DISCUSSION TITLE
JSON encoders are not respected when using a response_model

### DISCUSSION THREAD

**Discussion by pquadri** [fastapi/fastapi#8579] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import pydantic
from fastapi import FastAPI
from pydantic import BaseModel


class ORMConfig:
    orm_mode = True 


@pydantic.dataclasses.dataclass(config=ORMConfig)
class Money: 
    amount: float 
    currency: str 

    def __str__(self): 
        return f"{self.amount}/{self.currency}"


class PriceRange(BaseModel):
    min: Money 
    max: Money 

    class Config: 
        json_encoders = {Money: str}

price_range = PriceRange(min=Money(1000, 'EUR'), max=Money(2000, 'EUR'))

def create_app() -> FastAPI:

    app = FastAPI()


    @app.get("/json")
    def get_json(): 
        return price_range

    @app.get("/model", response_model=PriceRange)
    def get_model(): 
        return price_range

    return app
```

### Description

After the last release json encoders should be supported in fastapi responses, but when a response model is specified they are ignored.
You can check that the json endpoint yields 
`{"min":"1000.0/EUR","max":"2000.0/EUR"}` and the second one `{"min":{"amount":1000.0,"currency":"EUR"},"max":{"amount":2000.0,"currency":"EUR"}}`. 

### Operating System

Linux

### FastAPI Version

0.73.0

### Python Version

3.10.0

### Additional Context

I did some research, it appears that when a response model is specified the object is cloned, but the model config is not. Just adding the config to the new model is not enough since the types actually changed, so the configuration gets ignored.
I'm trying to figure out if / how to adapt the configuration to the newly built model (and, for my case in particular, the dataclass doesn't have its own __str__ method anymore).
Probably just giving the option of selecting whether to use the original field or clone it would be enough to get a workaround and leave the user the option to do it :)

**Comment by antmaiorino** [comment]
Would it be possible to have any update on this?

**Comment by arseniiarsenii** [comment]
I have a similar issue. Having the following response model:
```python
from decimal import Decimal

from pydantic import BaseModel


class SampleResponse(BaseModel):
    some_number: Decimal

    class Config:
        json_encoders = {Decimal: str}
```
I get field 'some_number' having type 'number' in the OpenAPI schema

**Comment by dolfinus** [comment]
`json_encoders` value can be any callable, not just some class constructor. How can FastAPI infer results type if this callable has generic type hints or no type hints at all?

**Comment by YuriiMotov** [comment, marked as answer]
This returns correct result with Pydantic V2.

In general, you should probably consider using separate model to define response model and define field validator for those fields. This will produce correct openapi schema:

```python
from typing import Annotated, Any, TypeAlias

import pydantic
from fastapi import FastAPI
from pydantic import BaseModel, ConfigDict, Field, field_validator


@pydantic.dataclasses.dataclass()
class Money:
    amount: float
    currency: str

    def __str__(self):
        return f"{self.amount}/{self.currency}"


class PriceRange(BaseModel):
    min: Money
    max: Money


MoneyStr: TypeAlias = Annotated[str, Field(examples=["1000/EUR", "2000/EUR"])]


class PriceRangeOutput(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    min: MoneyStr
    max: MoneyStr

    @field_validator("min", "max", mode="plain")
    @classmethod
    def stringify_money(cls, value: Any) -> str:
        if isinstance(value, Money):
            return str(value)
        elif isinstance(value, str):
            return value
        elif isinstance(value, dict) and "amount" in value and "currency" in value:
            return str(Money(**value))
        raise ValueError(f"Invalid value for Money: {value}")


price_range = PriceRange(min=Money(1000, "EUR"), max=Money(2000, "EUR"))


app = FastAPI()


@app.get("/json")
def get_json():
    return price_range


@app.get("/model", response_model=PriceRangeOutput)
def get_model():
    return price_range

```
